### PR TITLE
Modvend 81 - Load sample data via tenant API

### PR DESCRIPTION
**Pull request by komick**
## Purpose
In a recent discussion we've decided to pursue a different approach to loading sample data which involves overriding the default implementation of the tenant interface (POST _/tenant).

Jira
https://issues.folio.org/browse/MODVEND-81

## Approach
Since a similar process was performed on mod-orders-storage recently (https://issues.folio.org/browse/MODORDSTOR-48), the foundation of that solution was brought into mod-vendors.  Also, recent changes made to RMB 23.11.0 allowed for tenant data loading to be simplified.

The scope was to only change the method for loading current sample data records as they are, in order to replace the scripts used in the build environment.  As such, the vendor records have been left as monolithic files.  Breaking these into submodule-specific files will be completed as part of a future task (https://issues.folio.org/projects/MODVEND/issues/MODVEND-91).

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -94,6 +101,10 @@
       <version>1.0.4</version>
     </dependency>
 
+
```
Comment (craigmcnally) [inline pom.xml]: probably don't need these extra lines

File: src/main/java/org/folio/rest/impl/TenantReferenceAPI.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        hndlr.handle(res);
+        return;
+      }
+    }, cntxt);
+
+  }
+
+  private boolean isLoadSample(TenantAttributes tenantAttributes) {
+    // if a system parameter is passed from command line, ex: loadSample=true
+    // that
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/TenantReferenceAPI.java]: not
sure
what's
up
with
these
line
breaks

File: src/test/java/org/folio/rest/impl/CategoryTest.java [hunk 3]
```diff
@@ -101,7 +101,7 @@ private void verifyCollection() {
     getData("/vendor-storage/categories").then()
       .log().all()
       .statusCode(200)
-      .body("total_records", equalTo(4));
+      .body("total_records", equalTo(0));
```
Comment (craigmcnally) [inline src/test/java/org/folio/rest/impl/CategoryTest.java:102]: CPM errors in the comments above

File: src/test/java/org/folio/rest/impl/TenantSampleDataTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    finally {
+      deleteTenant(ANOTHER_TENANT_HEADER);
+    }
+  }
+
+  @Test
+  public void failIfNoUrlToHeader() throws MalformedURLException {
+    JsonObject jsonBody = TenantApiTestUtil.prepareTenantBody(true, false);
+    given()
+      .header(new Header(OKAPI_HEADER_TENANT, "noURL"))
```
Comment (craigmcnally) [inline src/test/java/org/folio/rest/impl/TenantSampleDataTest.java]: nitpick... flat indentation here isn't ideal

File: src/test/java/org/folio/rest/utils/TenantApiTestUtil.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  public static void prepareTenant(Header tenantHeader, boolean isLoadSampleData) throws MalformedURLException {
+    JsonObject jsonBody = prepareTenantBody(isLoadSampleData, false);
+    postToTenant(tenantHeader, jsonBody).statusCode(201);
+  }
+
+  public static ValidatableResponse postToTenant(Header tenantHeader, JsonObject jsonBody) throws MalformedURLException {
+    return given()
+      .header(tenantHeader)
```
Comment (craigmcnally) [inline src/test/java/org/folio/rest/utils/TenantApiTestUtil.java]: indentation again

**Review by craigmcnally — APPROVED** [review]
A few minor nitpicks but otherwise I think this looks good

File: src/main/resources/data/categories/Returns.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (javalkar-varun) [inline src/main/resources/data/categories/Returns.json]: Name of the file to be consistent with other filenames should start with lowercase `returns.json`

File: src/main/resources/data/categories/Shipments.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+ {
```
Comment (javalkar-varun) [inline src/main/resources/data/categories/Shipments.json]: Name of the file to be consistent with other filenames should start with lowercase letter`shipments.json`

**Comment by craigmcnally** [comment]
Is the category stuff considered sample or reference data?  Was it loaded via different mechanism than the vendor data because it's not sample, but reference data?

**Comment by KVupp** [comment]
> Is the category stuff considered sample or reference data? Was it loaded via different mechanism than the vendor data because it's not sample, but reference data?

yea I was wondering that too, but then previously it was loaded via schema.json along with other data. If we decide that it is reference data, it would just be another parameter loadReference and hence just another block of code.

**Comment by craigmcnally** [comment]
OK, Dennis and I have been working on a feature that will introduce some reference data, we can move the category data over to reference data as part of that work if needed

**Review by javalkar-varun — APPROVED** [review]

File: src/main/java/org/folio/rest/impl/TenantReferenceAPI.java (hunk 2 above)
Comment (KVupp) [reply src/main/java/org/folio/rest/impl/TenantReferenceAPI.java]: Done

File: src/test/java/org/folio/rest/impl/CategoryTest.java (hunk 3 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/CategoryTest.java:102]: Done

File: src/test/java/org/folio/rest/impl/TenantSampleDataTest.java (hunk 4 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/TenantSampleDataTest.java]: Done

File: src/test/java/org/folio/rest/utils/TenantApiTestUtil.java (hunk 5 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/utils/TenantApiTestUtil.java]: Done

File: src/main/resources/data/categories/Returns.json (hunk 6 above)
Comment (KVupp) [reply src/main/resources/data/categories/Returns.json]: Done

File: src/main/resources/data/categories/Shipments.json (hunk 7 above)
Comment (KVupp) [reply src/main/resources/data/categories/Shipments.json]: Done

File: pom.xml (hunk 1 above)
Comment (KVupp) [reply pom.xml]: Done

**Pull request merged**